### PR TITLE
Add commit ts to restore status

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb-operator/issues/2774.

### What is changed and how does it work?
Reuse the `getCommitTs` function for BR and lightning. Save commitTS in `restoreStatus`.

### Check List 

Tests 

 - Unit test
 - E2E test
 - Manual test (add detailed scripts or steps below)
The manual checks for lightning and br restore are both okay.
```yaml
  ...
    secretName: restore-demo-basic-secret
    user: root
status:
  commitTs: "417953765864243201"
  conditions:
  ...
```

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support show commitTS in restore status.
```

### CODE REVIEW

File: cmd/backup-manager/app/util/remote.go [hunk 1]
```diff
@@ -58,18 +58,18 @@ type gcsQuery struct {
 }
 
 // NewRemoteStorage creates new remote storage
-func NewRemoteStorage(backup *v1alpha1.Backup) (*blob.Bucket, error) {
-	st := util.GetStorageType(backup.Spec.StorageProvider)
+func NewRemoteStorage(provider v1alpha1.StorageProvider, fakeRegion bool) (*blob.Bucket, error) {
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/util/remote.go:61]: Why add `fakeRegion` to this function?

**Comment by lichunzhu** [comment]
@DanielZhangQD @cofyc PTAL again

File: cmd/backup-manager/app/import/manager.go [hunk 2]
```diff
@@ -186,6 +201,7 @@ func (rm *RestoreManager) performRestore(restore *v1alpha1.Restore) error {
 
 	restore.Status.TimeStarted = metav1.Time{Time: started}
 	restore.Status.TimeCompleted = metav1.Time{Time: finish}
+	restore.Status.CommitTs = commitTs
```
Comment (cofyc) [inline cmd/backup-manager/app/import/manager.go:204]: always update `commitTs` when we get it successfully

File: cmd/backup-manager/app/restore/manager.go [hunk 3]
```diff
@@ -247,9 +247,24 @@ func (rm *Manager) performRestore(restore *v1alpha1.Restore, db *sql.DB) error {
...
+			Message: err.Error(),
+		})
+		errs = append(errs, uerr)
+		return errorutils.NewAggregate(errs)
+	}
+
 	finish := time.Now()
 	restore.Status.TimeStarted = metav1.Time{Time: started}
 	restore.Status.TimeCompleted = metav1.Time{Time: finish}
+	restore.Status.CommitTs = fmt.Sprintf("%d", commitTs)
```
Comment (cofyc) [inline cmd/backup-manager/app/restore/manager.go:267]: nit: use `strconv.FormatInt(commitTs, 10)`

File: cmd/backup-manager/app/import/manager.go (hunk 2 above)
Comment (cofyc) [reply cmd/backup-manager/app/import/manager.go:204]: sorry, maybe it's better not to update it when the restore fails as the `commitTs` indicates the commit timestamp of the restored backup

File: cmd/backup-manager/app/util/util.go [hunk 4]
```diff
@@ -266,6 +271,74 @@ func GetOptions(provider v1alpha1.StorageProvider) []string {
...
+
+		Started dump at: 2019-06-13 10:00:04
+		SHOW MASTER STATUS:
+			Log: tidb-binlog
+			Pos: 409054741514944513
+			GTID:
+
+		Finished dump at: 2019-06-13 10:00:04
+*/
+func GetCommitTsFromMetadata(backupPath string) (string, error) {
```
Comment (cofyc) [inline cmd/backup-manager/app/util/util.go:287]: Can you add a unit test function for this function?

File: cmd/backup-manager/app/util/util.go [hunk 5]
```diff
@@ -266,6 +271,74 @@ func GetOptions(provider v1alpha1.StorageProvider) []string {
...
+			return commitTs, fmt.Errorf("parse mydumper's metadata file %s failed, str: %s", metaFile, lineStr)
+		}
+		commitTs = strings.TrimSpace(lineStrSlice[1])
+		break
+	}
+	return commitTs, nil
+}
+
+// GetCommitTsFromBRMetaData get backup position from `EndVersion` in BR backup meta
+func GetCommitTsFromBRMetaData(provider v1alpha1.StorageProvider) (uint64, error) {
```
Comment (cofyc) [inline cmd/backup-manager/app/util/util.go:314]: ditto, better to have a unit test for this function

File: cmd/backup-manager/app/import/manager.go (hunk 2 above)
Comment (lichunzhu) [reply cmd/backup-manager/app/import/manager.go:204]: which means current implementation is correct?

Comment (cofyc) [reply cmd/backup-manager/app/import/manager.go:204]: changes in this file LGTM

Comment (lichunzhu) [reply cmd/backup-manager/app/import/manager.go:204]: I see

File: cmd/backup-manager/app/restore/manager.go [hunk 6]
```diff
@@ -247,9 +247,24 @@ func (rm *Manager) performRestore(restore *v1alpha1.Restore, db *sql.DB) error {
...
+		errs = append(errs, err)
+		klog.Errorf("get cluster %s commitTs failed, err: %s", rm, err)
+		uerr := rm.StatusUpdater.Update(restore, &v1alpha1.RestoreCondition{
+			Type:    v1alpha1.RestoreFailed,
+			Status:  corev1.ConditionTrue,
+			Reason:  "GetCommitTsFailed",
+			Message: err.Error(),
+		})
+		errs = append(errs, uerr)
+		return errorutils.NewAggregate(errs)
```
[the remark this answers is not in the document]
Comment (cofyc) [reply cmd/backup-manager/app/restore/manager.go:261]: I think it's ok. I don't think it's necessary to continue the restore operation with the backup from which we can't even get the commitTs information.

File: cmd/backup-manager/app/util/util.go (hunk 4 above)
Comment (lichunzhu) [reply cmd/backup-manager/app/util/util.go:287]: added in https://github.com/pingcap/tidb-operator/pull/2899/commits/b6e31209b85f6ed44d286276031925c41a4be132

File: cmd/backup-manager/app/util/util.go (hunk 5 above)
Comment (lichunzhu) [reply cmd/backup-manager/app/util/util.go:314]: this function get TS from S3 and ts is decoded by `proto.Unmarshal`. Maybe later we can mock it in another PR. This PR won't add this.

File: cmd/backup-manager/app/restore/manager.go (hunk 3 above)
Comment (lichunzhu) [reply cmd/backup-manager/app/restore/manager.go]: revised.

File: cmd/backup-manager/app/export/manager.go [hunk 7]
```diff
@@ -299,7 +299,7 @@ func (bm *BackupManager) performBackup(backup *v1alpha1.Backup, db *sql.DB) erro
 	}
 	klog.Infof("get cluster %s archived backup file %s size %d success", bm, archiveBackupPath, size)
 
-	commitTs, err := getCommitTsFromMetadata(backupFullPath)
+	commitTs, err := util.GetCommitTsFromMetadata(backupFullPath)
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/export/manager.go:302]: Please wait for #2897 which removes the temp data after the compression.

File: cmd/backup-manager/app/import/manager.go [hunk 8]
```diff
@@ -182,10 +182,26 @@ func (rm *RestoreManager) performRestore(restore *v1alpha1.Restore) error {
 	}
 	klog.Infof("restore cluster %s from backup %s success", rm, rm.BackupPath)
 
+	commitTs, err := util.GetCommitTsFromMetadata(unarchiveDataPath)
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/import/manager.go:185]: As per discussion, please move this block after the uncompression code block.

File: cmd/backup-manager/app/restore/manager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply cmd/backup-manager/app/restore/manager.go:262]: OK, then i think we can move this block before the GC time setting code block.

**Comment by lichunzhu** [comment]
After manual test, br & export & import can record commitTs correctly. PTAL again @cofyc @DanielZhangQD

**Pull request merged**